File: samples/VB/Geocoder.vb
Imports System.Net
Imports System.Net.Http
Imports System.Web.Script.Serialization

Public Class Geocoder
    Private Const UrlTemplate As String = "http://api.mapserv.utah.gov/api/v1/geocode/{{street}}/{{zone}}"
    Private ReadOnly _apiKey As String

    Public Sub New(apiKey As String)
        _apiKey = apiKey
    End Sub

    Public Function Locate(street As String, zone As String, Optional options As Dictionary(Of String, Object) = Nothing) As Location

        Dim url = UrlTemplate.Replace("{{street}}", street).Replace("{{zone}}", zone)
        If options Is Nothing
            options = New Dictionary(Of String, Object)
        End If

        options.Add("apikey", _apiKey)

        url += "?" 

        For Each pair As KeyValuePair(Of String,Object) In options
            url += "&" + String.Concat(Uri.EscapeDataString(pair.Key), "=", Uri.EscapeDataString(pair.Value.ToString()))
        Next

        Dim client = New HttpClient()
        Dim response = client.GetAsync(url).Result

        Dim responseString = response.Content.ReadAsStringAsync().Result

        Dim serializer = New JavaScriptSerializer()

        Dim resultContainer = serializer.Deserialize (Of ResultContainer)(responseString)

        If response.StatusCode <> HttpStatusCode.OK Or resultContainer.Status <> HttpStatusCode.OK
            Console.WriteLine("{0} {1} was not found. {2}", street, zone, resultContainer.Message)

            Return Nothing
        End If

        Dim result = resultContainer.Result

        Console.WriteLine("match: {0} score [{1}]", result.Score, result.MatchAddress)

        Return result.Location
    End Function

End Class

File: samples/C#/Geocoder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using Nustache.Core;

namespace GeocodingSample
{
    public class Geocoding
    {
        private const string UrlTemplate = "http://api.mapserv.utah.gov/api/v1/geocode/{{street}}/{{zone}}";

        public Geocoding(string apiKey)
        {
            ApiKey = apiKey;
        }

        private string ApiKey { get; set; }

        public async Task<Location> Locate(string street, string zone, Dictionary<string, object> options = null)
        {
            var url = Render.StringToString(UrlTemplate, new
                {
                    Street = street,
                    Zone = zone
                });

            if (options == null)
                options = new Dictionary<string, object>();

            options.Add("apiKey", ApiKey);
            url += "?" + string.Join("&", options.Select(x => string.Concat(
                Uri.EscapeDataString(x.Key), "=",
                Uri.EscapeDataString(x.Value.ToString()))));

            var request = new HttpClient();

            var response = await request.GetAsync(url);

            var resultContainer = await response.Content.ReadAsAsync<ResultContainer<GeocodeResult>>();

            if (response.StatusCode != HttpStatusCode.OK || resultContainer.Status != (int) HttpStatusCode.OK)
            {
                Console.WriteLine("{0} {1} was not found. {2}", street, zone, resultContainer.Message);
                return null;
            }

            var result = resultContainer.Result;

            Console.WriteLine("match: {0} score [{1}]", result.Score, result.MatchAddress);

            return result.Location;
        }
    }
}